backend/src/postsDataSource.ts

```diff
@@ -15,36 +19,58 @@ export class PostsDataSource extends DataSource {
     return this.users
   }
 
+  existsUser (id) {
+    return !!this.users.find(user => user.id === id)
+  }
+
   getPosts () {
     return this.posts
   }
 
-  upvote (postId, voter) {
+  upvote (postId, userId) {
     const foundItem = this.getPosts().find(item => item.id === postId)
-    const foundUser = this.getUsers().find(user => user.name === voter.name)
+    const foundUser = this.getUsers().find(user => user.id === userId)
     if (foundItem && foundUser) {
-      if (!foundItem.voters.includes(foundUser.name)) {
+      if (!foundItem.voters.includes(foundUser.id)) {
         foundItem.votes += 1
-        foundItem.voters.push(foundUser.name)
+        foundItem.voters.push(foundUser.id)
       }
     }
     return foundItem
   }
 
-  addPost (newPost) {
-    const foundUser = this.getUsers().find(user => user.name === newPost.author.name)
+  addPost (newPost, userId) {
+    const foundUser = this.getUsers().find(user => user.id === userId)
     if (foundUser) {
       const item = { id: uuidv4(), title: newPost.title, votes: 0, voters: [], author: foundUser }
       foundUser.posts.push(item)
       this.posts.push(item)
       return item
-    } else {
-      const item = { id: uuidv4(), title: newPost.title, votes: 0, voters: [], author: {} }
-      const user = { name: newPost.author.name, posts: [item] }
-      item.author = user
-      this.users.push(user)
-      this.posts.push(item)
-      return item
     }
+    return null
+  }
+
+  login (email, password) {
+    const foundUser = this.getUsers().find(user => user.email === email)
+
+    if (foundUser) {
+      if (compareSync(password, foundUser.password)) {
+        return sign({ id: foundUser.id }, JWTSECRET, { algorithm: 'HS256' })
+      }
+      throw new AuthenticationError('Wrong email/password combination')
+    }
+  }
+
+  addUser (name, email, password) {
+    if (password.length < 8) {
+      throw new UserInputError('Password must have at least 8 characters')
+    }
+    const foundUser = this.getUsers().find(user => user.email === email)
+    if (foundUser) {
+      throw new ForbiddenError('Email already taken by another user')
```

Comment:
Why is the thrown error in the tests or the graphql playground "Not Authorized" even if we throw different errors here?

Reply:
I believe you must `allowExternalErrors: true` in `graphql-shield`.